### ISSUE TITLE
Run subsequents on success or failure

### ISSUE THREAD

**Issue by Ragnoroct** [casey/just#1582]
Relates to: https://github.com/casey/just/issues/819

It looks like `recipe_1: || recipe_after_1` wasn't implemented. I was wondering if a feature like that could be added?

My use case would be to have a common setup and cleanup recipe I can add to other recipes. It also allows the cleanup recipe to run regardless and avoid prefixing everything with a dash.

Amazing tool by the way :)

**Comment by ethack** [comment]
+1 on having subsequents that work as cleanup routines.

I found a workaround. It's fairly clean for a one-liner cleanup command, but anything longer and it could get really ugly.

```
test:
    #!/bin/bash
    trap 'echo; echo cleanup' EXIT
    echo start
    sleep 60
```

Running it and hitting `ctrl-c` during the sleep causes the cleanup command to be triggered.
```bash
❯ ./just test
start
^C
cleanup
```

**Comment by amarao** [comment]
Ideas from https://github.com/casey/just/issues/1902:

* `&` run subsequent recipe (right part) after left part succeed (after it)
* `||` will run subsequent recipe only if left part failed (after it)
* `;` will run subsequent recipe after left part without considering it status.

One additional request: if '||' or  ';' was executed, repeat the failure message for the left part as the last line of the output, and to return non-zero exit code. (e.g. even if cleanup success, recipe failed).

**Comment by VladimirMarkelov** [comment]
While using `&` and `||`  is straightforward, it has two issues: 1) inconsistency: one and two character 2) just already has `&&` and adding `&` is a possible source of confusion: I expect the same troubles as they have in `C/C++` with `=` and `==` in the `if` condition. Automatically people would write `&&` when they really want to use `&`. Though, I do not have a good proposal from my side.

**Comment by ufoscout** [comment]
What about some function style syntax instead? 
```
recipe_1: before_1 && on_success(after_on_success_1) on_failure(after_on_failure_1 after_on_failure_2) always(cleanup_resources)
```

**Comment by amarao** [comment]
It could be if designed from scratch.

The problem is backward compatibility, old recipes should not be broken, so every new syntax should be superset of the old.

**Comment by ufoscout** [comment]
@amarao
I agree, but if you consider `on_success` as the default case, then these two are equivalent:

- `recipe_1: before_1 && on_success(after_on_success_1)`
- `recipe_1: before_1 && after_on_success_1`

So the old behavior is maintained

**Comment by amarao** [comment]
Okay, so we get something like 'execution filter', available for extension in the future. Filter is applied to the recipe name, and can either allow execution of the recipe or skip it.

Filters:

* `on_success` runs recipe only if current recipe ended with success
* `on_failure` runs recipe only if current recipe ended with failure

So far so good, and I like how concise it is.

`always` is worrying. I will use 'failed' 'ok' names of recipes to indicate the order.

```
myrecipe: ok1 && on_success(failed1) && on_success(failed2) && on_failure(failed3) && always(ok2)
```

Should `failed2` run? By '&&' logic it shouldn't, because `failed1` failed.
If `failed2` does not run, why ok2 run?

on_success and on_failure checks the current recipe status, && chains them together, but always breaks 'filter' model.

It's not a filter.  What is it? If it's a function, what values does it check? Mind, we may have very odd combination of failed recipes including failures of nested cleanup targets, so we should be very strict here.

I definitively like on_success/on_failure, because they are local, but `always` make me feel I can't describe it's properties.

Can I bring back `;` idea?

**Comment by ufoscout** [comment]
The way I would read it:

```
myrecipe: ok1 && on_success(failed1) && on_success(failed2) && on_failure(failed3) && always(ok2)
                      |                     |                          |                 |
                      v                     |                          |                 v
                 run because previous       |                          |             run in any case
                 recipe succeeded           v                          |
                                        do not run because previous    |
                                        receipt failed                 v
                                                               run because previous
                                                               recipe failed to run
                                                               (it didn't run at all)
```

anyway I agree that it can be confusing

**Comment by amarao** [comment]
Let me play devil's advocate:

```
myrecipe: ok1 && always(failed1) && on_success(failed2) && always(ok2) && on_success(failed3) && on_failure(failed4)
```

What this line should means? (Error in interpretation? Something interesting?)

**Comment by ufoscout** [comment]
@amarao an error maybe?

**Comment by amarao** [comment]
@ufoscout, let's play it strict. Try to formulate the error message and I'll try to get around it to get to unsound state.

**Comment by casey** [comment]
I think the function like syntax is too different from the existing `&&` syntax.

`;` is commonly used to join two statements on a single line, which is useful functionality, so I'm not sure if we should use it for something else.

**Comment by giorgiga** [comment]
Maybe the current `()` syntax could be extended instead?

```
recipe: pre1 pre2 && post1 post2 (finally: recipe_that_cleans_up)
```

We could also say that the current `&&` syntax is syntactic sugar and that the above is equivalent to

```
recipe: (before:pre1) (before:pre2) (after:post1) (after:post2) (finally: recipe_that_cleans_up)
```

---

Regardless of syntax, things can get confusing really quickly.

Even if we suppose that just "finally" is added (similar to "always" above - to me, "on_error" doesn't seem like it would be used very often, and it could be implemented by saving some success state in the current recipe and reading it in another "always" one), there will be cases where the correct/expected behavior is not super-intuitive.

For example, if I run `just` with this justfile:

```just
@default: pre (finally: cleanup "default")
  echo default

@pre: (finally: cleanup "pre")
  exit 1

@cleanup msg:
  echo cleanup {{msg}}
```

Does it print
```
cleanup pre
[error message]
```
or
```
cleanup pre1
cleanup default
[error message]
```

And even without errors:

```just
@default: pre (finally: cleanup "default")
  echo default

@pre: (finally: cleanup "pre")
  echo pre

@cleanup msg:
  echo cleanup {{msg}}
```

Does it print
```
pre
cleanup pre
default
cleanup default
```
or
```
pre
default
cleanup pre
cleanup default
```

**Comment by starfishpatkhoo** [comment]
I'd like to throw a suggestion here for a different approach to this problem - add a `[catch(RECIPE)]` attribute. 

In the OP, a cleanup recipe is needed, even (especially) if something failed along the way. I have that same need too. Sometimes it is just to clean up temp files, or maybe to reset a testing environment. In today's modern world, I can even imagine calling an AI to investigate the error and print a summary as part of the `catch`.

While the suggestions above for recipe chaining is great, there are (kind-of) workarounds with some recipe engineering. However, the ability to run a specific recipe on failure, a `catch` after a `try`, is simply not possible currently. This is something we could not do with `make` either, hence always a `make clean` _BEFORE_ a `make`.

**Comment by amarao** [comment]
[starfishpatkhoo](https://github.com/starfishpatkhoo), do you want to put it on catching recipe or on the recipe which need to be catched? Can you show an example?

**Comment by starfishpatkhoo** [comment]
Oh definitely `catch` on the recipe that needs to be caught, not on the "catcher" recipe. Not every recipe needs to be caught.

FYI: Apologies.. I have to summarise/edit the examples below because I cannot post an actual recipes, but hope you catch the idea

### Example 1: Process data

```justfile
[catch("_cleanup SOURCE_DIR")]
up_vrl SOURCE_DIR VRL FIXED_TERM: \
	( ask_vrl SOURCE_DIR ) \
	( _build_contact_url SOURCE_DIR VRL ) \
	( _upload_vrl SOURCE_DIR FIXED_TERM ) \
	( _link_vrl SOURCE_DIR VRL )

_cleanup SOURCE_DIR:
	set -eu
	msg_header "Cleanup '{{SOURCE_DIR}}'"
	...
```

In this example, `ask_vrl` would set some parameters and options inside the `SOURCE_DIR`. Then, other processing steps would occur. If, something went wrong in the other recipes, the `_cleanup` could reset the `SOURCE_DIR` back to a known state, so that I can run `up_vrl` again safely and quickly.


### Example 2: Use a temporary build folder

```justfile
[catch("_clear DIR_SITE")]
do DIR_SITE SOURCE_FILES: \
	_setup DIR_SITE \
	_build DIR_SITE SOURCE_FILES \
	_deploy DIR_SITE \
	_cleanup DIR_SITE

_clear DIR_SITE:
	set -eu
	msg_header "Clear '{{DIR_SITE}}'"
	rm -rf {{mp_path(DIR_SITE)}}
```

Since the folder is temporary, if something went wrong, I'd like it cleaned up anyway, even if stuff didn't work. Similar to this example is populating a test DB with junk data for testing. 


### Example 3: Testing

```justfile
[catch("_reset_db DB_TABLE")]
api_test DB_TABLE: \
	_populate_db DB_TABLE \
	_test_read DB_TABLE \
	_test_write DB_TABLE \
	_test_update DB_TABLE \
	_test_del DB_TABLE \
	_cleanup DB_TABLE

_reset_db DB_TABLE:
	set -eu
	sql drop {{DB_TABLE}}
```

I'm sure there are some other examples for docker, to either rm or clear cache or whatever, when things have gone wrong. And so on...


### Advanced Example 4: Force Require A Directory to Exist

```justfile
do_stuff PATH: \
	( _require_path_exist PATH )
	ls -l {{PATH}}

[catch("_create_path PATH")]
_require_path_exist PATH:
	{{assert("true" == path_exists(PATH), "Path Does Not Exist!\n")}}

_create_path PATH:
	mkdir -p {{PATH}}
```

And this is probably a very powerful example of what might be possible. It assumes that if the `catch` recipe returns "success", then the caught recipe will be treated as successful also.

- `do_stuff` needs `PATH` to exist. 
- If `PATH` does not exist (from the `assert`), then the `catch` will create it.
- If `_create_path` succeeded, then the recipe `_require_path_exist` in `do_stuff` was successful, and so the `ls -l` would execute
- If `_create_path` failed, then the recipe `_require_path_exist` in `do_stuff` is considered to have failed, and so `do_stuff` would stop right there (as per normal recipe execution).

**Comment by amarao** [comment]
It look interesting, but:

1. It should be composable. Can I have two cleanups for one recipe?
2. It should support nesting. Cleanup of cleanup (no, I'm not inventing, I already have this in my CIs, mostly with matrix stuff).
3. What happens if cleanup fails but we have cleanup? Should we report error?
4. How exactly error is preserved? If we have recipe which fails, then we have a cleanup, which succeed, should we report error? How user can understand where error happens? (e.g. 10 pages up on the screen).
5. It looks like a dependency tree to me, where we have conditions for following dependencies (always before, always after, on success after, on failure after). If we do DAG, maybe it's better to encode it as a proper dag.

This topic is really close to the way CI jobs are described in most CI servers.

**Comment by starfishpatkhoo** [comment]
>     1. It should be composable. Can I have two cleanups for one recipe?

I'm thinking no.. Hence my suggestion to use an attribute. Only one `catch` per `try` (recipe).

>     2. It should support nesting. Cleanup of cleanup (no, I'm not inventing, I already have this in my CIs, mostly with matrix stuff).

Yes, it should. If you `catch` down below in the stack, and that fails, then the catch is propogated up the stack. The next higher recipe with a `catch` will then catch it.

>     3. What happens if cleanup fails but we have cleanup? Should we report error?

Just propogate the error up the stack. Let parent recipes handle it, or `catch` it if they have a `catch` defined. Might this end up in some kind of infinite recursion loop? Yes. But that's clearly the programmer's choice, not the fault of just. You can write an infinite loop try/catch block in all of the languages with try/catch support.

>     4. How exactly error is preserved? If we have recipe which fails, then we have a cleanup, which succeed, should we report error? How user can understand where error happens? (e.g. 10 pages up on the screen).

Print the error like it is currently, and propogate / bubble the error up the stack. And yes, maybe that is 10 pages up the screen. If one wants a "panic" exit, don't use a `catch` - which is just's current behaviour anyway.

But I really think that if someone is using try/catch blocks, they ought to be coding recipes that log the error or print the error in a sane and obvious way. And part of the purpose of a `catch` is to handle the error properly. Maybe perform a retry. Or do some mitigation. Or log/print a special error message or debug dump. Or... Aside from simply cleaning up junk. This is basically the same design/consideration as try/catch in all the other languages too, so programmers can follow the same strategies.

>     5. It looks like a dependency tree to me, where we have conditions for following dependencies (always before, always after, on success after, on failure after). If we do DAG, maybe it's better to encode it as a proper dag.

I'm not sure about a dependency tree - it is not meant to be. I'm not suggesting the complexity of adding `and`, `or`, `on_success` or `on_failure` as flow-control or branching structures for a dependency or decision tree of nested recipes.

Instead, I'm thinking more along the lines of a simple exception handling stack. Again, I am modelling it on try/catch blocks in Javascript, C, PHP, Powershell, python, swift, kotlin, ruby, etc etc etc... I was surprised to read that even Nushell has try/catch too.. huh.

The way it works now, every recipe in just and make is basically a `try` block. If an error occurs somewhere, an exception is raised. If the error occurs in a nested recipe, the exception bubbles all the way to the top. The current default `catch` exception handler is to simply return the error up the stack.

What I am suggesting is to use an attribute to define some other recipe as the `catch`. If this other "catcher" recipe returns an error, then it just returns the error up the stack like usual. If however, the "catcher" recipe returns success, that overrides the original error and the recipe is treated as a success (even though it never actually executed all the way).

If we _really_ want to extend it further, then we can add the `finally` as yet another attribute to round out the try/catch/finally block.

> This topic is really close to the way CI jobs are described in most CI servers.

I agree. Don't we want CI jobs to do something special if and when an error occurs during the job? That's "try this set of instructions, and if something goes wrong somewhere, catch the error, and handle it like this..." ^_^

**Comment by amarao** [comment]
It's usually not 'error handling case', but cleanup. In 90% of CI jobs I wrote, it's not `if: failure`, but `if: always()`. The tricky part is ordering.

imagine this:

```
level2: level3
   ...

level3:
    ...

[catch("_cleanup")]
level1: level2 && next_recipe
   ...
```

and level3 fails. Is it 'catch', or not? For me it's completely not obvious, that calling level2 is not in 'catch' block.

(same question for `next_recipe`).

**Comment by starfishpatkhoo** [comment]
I think that when designing the catch blocks, they should be tied to the recipe that you are catching. Clean up only that recipe. Let other recipe's catch blocks clean up those recipes. But of course, when trying to understand the layers of a stack, it is also important to name the recipes (like variables) in a way that tells us what they are. Like I would name it `_cleanup_level1`, so I know that that recipe is the cleanup for `level1`. That helps also. But these are programmer design decisions, not really tied to the syntax and features of justfiles. I've added some references at the bottom for how other programmers do it for more ideas. I find it interesting to see the differences and similarities across the different languages for essentially a very similar code flow.

---

Consider the following examples... You probably want Example 4, but Examples 1, 2 and 3 will help explain Example 4 better.

### Example 1

```
parent: child
   ...

child: grandchild
   ...

[catch("catch_handler")]
grandchild:
   ...

catch_handler:
   ...
```

- If an error occurred in `grandchild`, then `catch_handler` will execute and `catch_handler` success/fail status will be used instead of `grandchild` status
- If an error occurred in `child` or `grandchild`, then the error bubbles up / is propagated up to `parent` (the default just handling)
- If an error occurred in `parent`, then all execution stops (the default just handling)
- TL;DR: If an error occurs in `grandchild`, `catch_handler` will be called, but if an error occurs in `child` or `parent`, then execution will stop (the default just handling)


### Example 2

```
[catch("catch_handler")]
parent: child
   ...

child: grandchild
   ...

grandchild:
   ...

catch_handler:
   ...
```

- If an error occurred in `grandchild`, then the error bubbles up / is propagated up to `child` (the default just handling)
- If an error occurred in `child`, then the error bubbles up / is propagated up to `parent` (the default just handling)
- If an error occurred in `parent`, then `catch_handler` will execute and `catch_handler` success/fail status will be used instead of `parent` status
- TL;DR: If an error occurs anywhere in `parent`, `child`, or `grandchild`, `catch_handler` will be called


### Example 3

```
[catch("catch_handler_A")]
parent: child
   ...

child: grandchild
   ...

[catch("catch_handler_B")]
grandchild:
   ...

catch_handler_A:
   ...

catch_handler_B:
   ...
```

- If an error occurred in `grandchild`, then `catch_handler_B` will execute and `catch_handler_B` success/fail status will be used instead of `grandchild` status
- If an error occurred in `child`, then the error bubbles up / is propagated up to `parent` (the default just handling)
- If an error occurred in `parent`, then `catch_handler_A`  will execute and `catch_handler_A` success/fail status will be used instead of `parent` status
- TL;DR: If an error occurs in `grandchild`, `catch_handler_B` will execute, but if an error occurs in `parent` or `child`, then `catch_handler_A` will execute. Of course, if an error occurred in `grandchild`, and `catch_handler_B` also returned an error, then `catch_handler_A` will also execute, because the error from `catch_handler_B` will be propagated up the stack, and `parent` `catch` will then catch it.


### Example 4: finally() or always()

Normally, we need to add some recipe to clean things up. It usually looks something like this:

```
parent: child && cleanup
   ...

child: grandchild
   ...

grandchild:
   ...

cleanup:
   ...
```

But what if an error occurs, and we still want `cleanup` to execute?

```
parent: real_parent && cleanup

[catch("catch_handler")]
real_parent: child
   ...

child: grandchild
   ...

grandchild:
   ...

cleanup:
   ...

catch_handler: cleanup
   exit 1
```

- If everything is smooth, `cleanup` will execute as per normal
- If an error occurred in `real_parent`, `child` or `grandchild`, `catch_handler` will be executed as part of the `catch`
- Notice that `catch_handler` will call `cleanup`, but will explicitly return an error. Therefore, `parent` will always see this error from `real_parent`, and the `&& cleanup` will not execute (a second time)
- Also, because we add the `catch` to `real_parent` and not `parent`, we will not have a loop of `cleanup` triggering `catch_handler` + `cleanup` if `cleanup` itself had an error

----

**TL;DR:** Ultimately, I would like a feature that yes, allows me to always clean up after, and/or handle errors that happen during. And I feel a try/catch mechanism using a recipe attribute:
- Is easier to implement in just than trying to implement a full decision/dependency tree or more advanced flow control mechanisms and special options like `on_success()` and `on_failure()`. 
- Is closer to how just (and make) already function, with their "fast fail" operation. 
- Stays out the the way of people who do not need it, thereby maintaining full backward compatibility, without worrying about some justfile somewhere messing it up. 
- Will also allow us (with some recipe engineering) to implement a safe and effective "finally()" or "always()" like in Example 4.

----


#### References

I admit, my explanation is poor... There are plenty of more advanced and detailed examples for the various languages out there... Of course, they have very advanced try/catch structures, with decades of experience behind them. I'm only suggesting the very basic of the basic. Catch an error from below, and execute catch code.

- C++: https://coddy.tech/learn/cpp/logic_and_flow/the_try_and_catch_blocks
- PHP: https://kanishkame.medium.com/mastering-try-catch-in-php-a-complete-guide-to-exception-handling-50bbfb36bebc
- Javascript: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/try...catch
- Python: https://www.w3schools.com/python/python_try_except.asp

**Comment by amarao** [comment]
In my opinion, restricting it only to catching errors is reducing composability. A good feature should be universal, and give users more than envisioned by the designer. Job dependencies (including failure handling) is a universal feature, 'catch error' is much more narrow and limiting.

**Comment by starfishpatkhoo** [comment]
> In my opinion, restricting it only to catching errors is reducing composability. A good feature should be universal, and give users more than envisioned by the designer. Job dependencies (including failure handling) is a universal feature, 'catch error' is much more narrow and limiting.

Well... Yes, you are right, especially if you consider some other suggested patterns above - like:

```justfile
recipe_1: before_1 && on_success(after_on_success_1) on_failure(after_on_failure_1 after_on_failure_2) always(cleanup_resources)
```

is much broader as you suggest. It is venturing into much more complex flow control, approaching - as you mentioned - dependency/decision trees with associated graphs... But it is complex because if you consider `on_failure(after_on_failure_1 after_on_failure_2)`, how would one add recipe arguments? Is `after_on_failure_2` an argument of `after_on_failure_1`? Recipe arguments are currently specified with brackets like `( recipe PARAM_1 PARAM_2 )`. 

In contrast, I am suggesting a narrower, but simpler implementation within just. Just recipes **ALREADY** have basic dependencies, both run before and run after. However, the only failure/error handling right now, is "fast exit". Implementing a `catch` attribute is merely to allow for a "custom" recipe failure/error handling/recovery that is fully nested-aware.

While it is a narrower implementation, I think it is quite powerful, and can still achieve what was intended:

```justfile
recipe_1: real_recipe_1 && do_on_success

[catch("do_on_failure")]
real_recipe_1: before_1

do_on_failure: after_on_failure_1 after_on_failure_2 cleanup_resources
  exit 1

do_on_success: after_on_success_1 cleanup_resources
```

- Notice that the Github code highlighting for justfile works just fine in the `catch` attribute example, but gets confused in the first one. So, as I mentioned, backward compatibility is greatly supported.
- Try/catch concepts are very well established with decades of past experience across many different languages - no need to reinvent the wheel so much when it comes to recipe engineering
- By using a recipe attribute, no new symbols like `&` or `||` or `on_failure()` or such needs to be added to the parsing of the recipe line, which might conflict with current symbols - https://github.com/casey/just/issues/1582#issuecomment-2405914020
- Attributes already have the capability to support variadic parameters (see `[arg()]`), and even expressions (see `[env()]`). All this stuff is "within reach" for a `[catch()]` implementation - but if we build something like `on_failure()`, we'd have to start all over again for those things.

Anyways, I've given plenty of recipe examples how a try/catch approach can be used to achieve different objectives, along with multiple reasons why I feel this is a better choice (not that implementing a full dependency tree is a bad choice). 

I guess in the end, it is a design decision for just and @casey , and I am fine with anything, as long as I can have more flexible recipes to handle different situations. ^_^
